host-controller/src/main/java/org/jboss/as/domain/controller/operations/SyncServerStateOperationHandler.java

```diff
@@ -126,30 +148,51 @@ public void handleResult(OperationContext.ResultAction resultAction, OperationCo
+        final Map<String, SyncServerResultAction> serverStateChanges = new HashMap<>();
         for (ModelNode operation : operations) {
-            PathAddress addr = PathAddress.pathAddress(operation.get(OP_ADDR));
+            final PathAddress addr = PathAddress.pathAddress(operation.get(OP_ADDR));
             contentDownloader.checkContent(operation, addr);
-
-            Map<Set<ServerIdentity>, ModelNode> serverMap = resolver.getServerOperations(context, operation, addr);
+            final ImmutableManagementResourceRegistration registration = context.getResourceRegistration();
+            final Map<Set<ServerIdentity>, ModelNode> serverMap = resolver.getServerOperations(context, operation, addr);
             for (Map.Entry<Set<ServerIdentity>, ModelNode> entry : serverMap.entrySet()) {
-                ModelNode op = entry.getValue();
-                boolean restart = op.get(OP).asString().equals(ServerProcessStateHandler.REQUIRE_RESTART_OPERATION);
-                for (ServerIdentity id : entry.getKey()) {
-                    String serverName = id.getServerName();
-                    Boolean existing = serverStateChanges.get(serverName);
-                    if (existing == null || (!existing && restart)) {
-                        serverStateChanges.put(serverName, restart);
+                final ModelNode op = entry.getValue();
+                final String opName = op.get(OP).asString();
+
+                // default is RELOAD_REQUIRED
+                SyncServerResultAction action = SyncServerResultAction.RELOAD_REQUIRED;
+                if (WRITE_ATTRIBUTE_OPERATION.equals(opName) || UNDEFINE_ATTRIBUTE_OPERATION.equals(opName)) {
+                    // look up the attribute name we're writing, and check the flags to see if we need restart.
+                    final String attributeName = op.get(NAME).asString();
+                    // look up if the attribute requires restart / reload
+                    final EnumSet<AttributeAccess.Flag> flags = registration.getAttributeAccess(addr, attributeName).getAttributeDefinition().getFlags();
+                    if (flags.contains(AttributeAccess.Flag.RESTART_JVM)) {
+                        action = SyncServerResultAction.RESTART_REQUIRED;
+                    }
+                } else { // all other ops
+                    final Set<OperationEntry.Flag> flags = registration.getOperationFlags(addr, opName);
+                    if (flags.contains(OperationEntry.Flag.RESTART_JVM)) {
+                        action = SyncServerResultAction.RESTART_REQUIRED;
                     }
                 }
+
+                for (final ServerIdentity id : entry.getKey()) {
+                    final String serverName = id.getServerName();
+                    serverStateChanges.put(serverName, serverStateChanges.get(serverName) == null ? SyncServerResultAction.RELOAD_REQUIRED : action);
```

Comment:
Yah, sorry about those, I'd been going back and forth trying to avoid really verbose checking, will make that clear & fix.